packages/php-wasm/logger/src/lib/logger.ts

```diff
@@ -256,77 +135,125 @@ export class Logger extends EventTarget {
- * @param loggerInstance The logger instance
- */
-export function collectWindowErrors(loggerInstance: Logger) {
-	loggerInstance.addWindowErrorListener();
-	loggerInstance.addServiceWorkerMessageListener();
-}
+	/**
+	 * Log message with severity.
+	 *
+	 * @param message any
+	 * @param severity LogSeverity
+	 * @param args any
+	 */
+	private logMessage(
+		message: any,
+		severity?: LogSeverity,
+		...args: any[]
+	): void {
+		this.addLogEntry(this.prepareLogMessage(message, ...args), severity);
+		this.consoleLog(message, severity, ...args);
+	}
 
-/**
- * Collect PHP logs from the error_log file and log them.
- * @param UniversalPHP playground instance
- * @param loggerInstance The logger instance
- */
-export function collectPhpLogs(
-	loggerInstance: Logger,
-	playground: UniversalPHP
-) {
-	loggerInstance.addPlaygroundRequestEndListener(playground);
-}
+	/**
+	 * Log message
+	 *
+	 * @param message any
+	 * @param args any
+	 */
+	public log(message: any, ...args: any[]): void {
+		this.logMessage(message, undefined, ...args);
+	}
 
-/**
- * **Call this inside a service worker.**
- *
- * Reports service worker metrics.
- * Allows the logger to request metrics from the service worker by sending a message.
- * The service worker will respond with the number of open Playground tabs.
- *
- * @param worker The service worker
- */
-export function reportServiceWorkerMetrics(worker: ServiceWorkerGlobalScope) {
-	worker.addEventListener('activate', () => {
-		worker.clients.matchAll().then((clients) => {
-			const metrics = {
-				numberOfOpenPlaygroundTabs: clients.filter(
-					// Only count top-level frames to get the number of tabs.
-					(c) => c.frameType === 'top-level'
-				).length,
-			};
-			worker.clients
-				.matchAll({
-					// We don't claim the clients, so we need to include uncontrolled clients to inform all tabs.
-					includeUncontrolled: true,
-				})
-				.then((clients) => {
-					for (const client of clients) {
-						client.postMessage(metrics);
-					}
-				});
-		});
-	});
+	/**
+	 * Log debug message
+	 *
+	 * @param message any
+	 * @param args any
+	 */
+	public debug(message: any, ...args: any[]): void {
+		this.logMessage(message, 'Debug', ...args);
+	}
+
+	/**
+	 * Log info message
+	 *
+	 * @param message any
+	 * @param args any
+	 */
+	public info(message: any, ...args: any[]): void {
+		this.logMessage(message, 'Info', ...args);
+	}
+
+	/**
+	 * Log warning message
+	 *
+	 * @param message any
+	 * @param args any
+	 */
+	public warn(message: any, ...args: any[]): void {
+		this.logMessage(message, 'Warn', ...args);
+	}
+
+	/**
+	 * Log error message
+	 *
+	 * @param message any
+	 * @param args any
+	 */
+	public error(message: any, ...args: any[]): void {
+		this.logMessage(message, 'Error', ...args);
+	}
+
+	/**
+	 * Log message to the console.
+	 * @param string log
+	 * @param LogSeverity severity
+	 * @param any args
+	 * @returns void
+	 */
+	public consoleLog(
```

Comment:
Rather than a special case and a separate method call, this could be one of the "log handlers" / "data sinks" for all the messages that's no different than logging to a file or a REST API.

Reply:
I added support for log handlers in [a5c9ec1](https://github.com/WordPress/wordpress-playground/pull/1264/commits/a5c9ec1de88675652ae190682e04581e13ff7cde).

The current implementation has a few TODOs because the logToFile handler still uses an array internally. I will address this separately to keep this PR focused on refactoring.